# ForecastLab

---

## Bike Sharing Demand Forecasting

---

| Informasi | Detail |
|-----------|--------|
| **Project ID** | FL-001 |
| **Eksperimen** | EXP-001 |
| **Nama Eksperimen** | Data Understanding |
| **Dataset** | hour.csv (Bike Sharing Dataset — UCI ML Repository) |
| **Tujuan Notebook** | Memahami struktur, kualitas, dan karakteristik awal dataset |
| **Tanggal Pengerjaan** | 21 Juli 2026 |
| **Versi Notebook** | 1.0.0 |
| **Author** | Najib Bahrudin |

---

## Daftar Isi

1. [Tujuan Eksperimen](#1-tujuan-eksperimen)
2. [Import Library](#2-import-library)
3. [Load Dataset](#3-load-dataset)
4. [Dataset Overview](#4-dataset-overview)
5. [Data Structure](#5-data-structure)
6. [Missing Value Analysis](#6-missing-value-analysis)
7. [Duplicate Analysis](#7-duplicate-analysis)
8. [Statistical Summary](#8-statistical-summary)
9. [Validasi Data Dictionary](#9-validasi-data-dictionary)
10. [Initial Findings](#10-initial-findings)
11. [Next Step](#11-next-step)

---

## 1. Tujuan Eksperimen

### Tujuan

Memahami struktur dataset, tipe data, kualitas data, serta karakteristik awal dataset sebelum dilakukan analisis lebih lanjut.

Tahapan ini merupakan langkah pertama yang wajib diselesaikan sebelum masuk ke tahap Exploratory Data Analysis (EDA) sesuai aturan eksperimen pada Experiment Plan.

---

### Hipotesis

> Dataset memiliki struktur yang lengkap, tidak terdapat missing value yang signifikan, dan seluruh feature sesuai dengan dokumentasi pada Data Dictionary.

---

### Output yang Diharapkan

- Ringkasan dimensi dan struktur dataset
- Statistik deskriptif seluruh kolom
- Laporan missing value dan duplicate
- Validasi kesesuaian dataset dengan Data Dictionary
- Catatan awal (initial findings) sebagai dasar EDA

---

### Deliverable

- ✅ Notebook selesai dan dapat dijalankan ulang (reproducible)
- ✅ Dataset tervalidasi terhadap Data Dictionary
- ✅ Insight awal didokumentasikan

---

### Acceptance Criteria

| Kriteria | Status |
|----------|--------|
| Dataset berhasil dimuat | Diperiksa pada Section 3 |
| Jumlah record sesuai dokumentasi (17.379) | Diperiksa pada Section 4 |
| Jumlah kolom sesuai dokumentasi (17) | Diperiksa pada Section 4 |
| Missing value telah diperiksa | Diperiksa pada Section 6 |
| Duplicate telah diperiksa | Diperiksa pada Section 7 |
| Target prediksi telah dipastikan | Diperiksa pada Section 9 |

---

## 2. Import Library

Pada tahap Data Understanding, library yang dibutuhkan adalah:

- **pandas** — untuk manipulasi dan analisis data tabular
- **numpy** — untuk operasi numerik

Library visualisasi belum diperlukan pada tahap ini dan akan diimport pada notebook EDA.

In [1]:
import pandas as pd
import numpy as np

# Konfigurasi tampilan pandas
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:.4f}')

print('Library berhasil diimport.')
print(f'Versi pandas : {pd.__version__}')
print(f'Versi numpy  : {np.__version__}')

Library berhasil diimport.
Versi pandas : 2.3.3
Versi numpy  : 2.0.2


---

## 3. Load Dataset

Dataset yang digunakan pada proyek ini adalah `hour.csv` yang berisi data penyewaan sepeda per jam dari sistem Capital Bikeshare, Washington D.C., USA.

Dataset didefinisikan menggunakan variabel `DATA_PATH` agar mudah diperbarui jika diperlukan.

In [2]:
# Definisikan path dataset
DATA_PATH = '../data/hour.csv'

# Load dataset
df = pd.read_csv(DATA_PATH)

print(f'Dataset berhasil dimuat dari: {DATA_PATH}')
print(f'Jumlah baris  : {df.shape[0]:,}')
print(f'Jumlah kolom  : {df.shape[1]}')

Dataset berhasil dimuat dari: ../data/hour.csv
Jumlah baris  : 17,379
Jumlah kolom  : 17


**Interpretasi:**

Dataset berhasil dimuat. Langkah selanjutnya adalah melakukan pemeriksaan awal terhadap konten dataset untuk memastikan data terbaca dengan benar.

---

## 4. Dataset Overview

Pada bagian ini dilakukan pemeriksaan awal untuk mendapatkan gambaran umum dataset.

### 4.1 Dimensi Dataset

Menampilkan jumlah baris dan kolom dataset.

In [3]:
# Dimensi dataset
baris, kolom = df.shape

print('=' * 40)
print('DIMENSI DATASET')
print('=' * 40)
print(f'Jumlah Baris  : {baris:,}')
print(f'Jumlah Kolom  : {kolom}')

DIMENSI DATASET
Jumlah Baris  : 17,379
Jumlah Kolom  : 17


**Interpretasi:**

Dataset memiliki **17.379 baris** dan **17 kolom**, sesuai dengan dokumentasi pada Data Dictionary. Setiap baris merepresentasikan kondisi penyewaan sepeda pada satu jam tertentu selama dua tahun (2011–2012).

### 4.2 Nama Kolom

Menampilkan seluruh nama kolom yang tersedia pada dataset.

In [4]:
# Menampilkan nama kolom
print('=' * 40)
print('NAMA KOLOM DATASET')
print('=' * 40)
for i, col in enumerate(df.columns, 1):
    print(f'{i:>2}. {col}')

NAMA KOLOM DATASET
 1. instant
 2. dteday
 3. season
 4. yr
 5. mnth
 6. hr
 7. holiday
 8. weekday
 9. workingday
10. weathersit
11. temp
12. atemp
13. hum
14. windspeed
15. casual
16. registered
17. cnt


**Interpretasi:**

Terdapat 17 kolom yang terdiri dari identifier, time features, calendar features, weather features, user features, dan target prediksi. Penamaan kolom terlihat konsisten dengan dokumentasi Data Dictionary.

### 4.3 Lima Sampel Data Pertama

Menampilkan lima baris pertama untuk melihat isi data.

In [5]:
# Lima baris pertama
df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.2400,0.2879,0.8100,0.0000,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.2200,0.2727,0.8000,0.0000,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.2200,0.2727,0.8000,0.0000,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.2400,0.2879,0.7500,0.0000,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.2400,0.2879,0.7500,0.0000,0,1,1


**Interpretasi:**

Lima baris pertama menunjukkan data dimulai dari **1 Januari 2011 pukul 00:00** (jam ke-0). Kolom `instant` merupakan index sekuensial, `dteday` adalah tanggal, dan kolom lainnya merepresentasikan kondisi waktu, cuaca, serta jumlah penyewaan.

### 4.4 Lima Sampel Data Terakhir

Menampilkan lima baris terakhir untuk memastikan rentang data.

In [6]:
# Lima baris terakhir
df.tail()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
17374,17375,2012-12-31,1,1,12,19,0,1,1,2,0.2600,0.2576,0.6000,0.1642,11,108,119
17375,17376,2012-12-31,1,1,12,20,0,1,1,2,0.2600,0.2576,0.6000,0.1642,8,81,89
17376,17377,2012-12-31,1,1,12,21,0,1,1,1,0.2600,0.2576,0.6000,0.1642,7,83,90
17377,17378,2012-12-31,1,1,12,22,0,1,1,1,0.2600,0.2727,0.5600,0.1343,13,48,61
17378,17379,2012-12-31,1,1,12,23,0,1,1,1,0.2600,0.2727,0.6500,0.1343,12,37,49


**Interpretasi:**

Data berakhir pada **31 Desember 2012**, mengonfirmasi bahwa dataset mencakup dua tahun penuh yaitu 2011 dan 2012. Dataset bersifat time series yang berurutan secara kronologis.

### 4.5 Sampel Data Acak

Menampilkan 5 baris acak untuk mendapatkan gambaran variasi data.

In [7]:
# Sampel data acak (random state untuk reproducibility)
df.sample(5, random_state=42)

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
12830,12831,2012-06-23,3,1,6,19,0,6,0,1,0.8000,0.6970,0.2700,0.1940,185,240,425
8688,8689,2012-01-02,1,1,1,20,1,1,0,1,0.2400,0.2273,0.4100,0.2239,5,83,88
7091,7092,2011-10-28,4,0,10,2,0,5,1,1,0.3200,0.3030,0.6600,0.2836,1,3,4
12230,12231,2012-05-29,2,1,5,19,0,2,1,1,0.7800,0.7121,0.5200,0.3582,69,457,526
431,432,2011-01-20,1,0,1,0,0,4,1,1,0.2600,0.2273,0.5600,0.3881,5,8,13


**Interpretasi:**

Sampel acak menunjukkan variasi data pada berbagai kondisi waktu dan cuaca. Kolom `cnt` (total penyewaan) tampak bervariasi sesuai dengan kondisi masing-masing baris.

---

## 5. Data Structure

Pada bagian ini dilakukan analisis terhadap struktur internal dataset, termasuk tipe data setiap kolom dan kategorisasi feature sesuai Data Dictionary.

### 5.1 Informasi Dataset (`info()`)

Menampilkan ringkasan lengkap struktur dataset.

In [8]:
# Informasi umum dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


**Interpretasi:**

- Tidak ada nilai null pada seluruh kolom (semua non-null count sama dengan 17.379).
- Kolom `dteday` bertipe `object` (string), perlu dikonversi ke tipe `datetime` pada tahap preprocessing.
- Kolom `temp`, `atemp`, `hum`, dan `windspeed` bertipe `float64`.
- Seluruh kolom lainnya bertipe `int64`.

### 5.2 Tipe Data per Kolom

In [9]:
# Tipe data setiap kolom
print('=' * 50)
print('TIPE DATA SETIAP KOLOM')
print('=' * 50)
dtype_df = pd.DataFrame({
    'Kolom': df.dtypes.index,
    'Tipe Data': df.dtypes.values
}).reset_index(drop=True)
print(dtype_df.to_string(index=False))

TIPE DATA SETIAP KOLOM
     Kolom Tipe Data
   instant     int64
    dteday    object
    season     int64
        yr     int64
      mnth     int64
        hr     int64
   holiday     int64
   weekday     int64
workingday     int64
weathersit     int64
      temp   float64
     atemp   float64
       hum   float64
 windspeed   float64
    casual     int64
registered     int64
       cnt     int64


### 5.3 Kategori Feature

Mengklasifikasikan seluruh kolom sesuai dengan Data Dictionary.

In [10]:
# Klasifikasi feature sesuai Data Dictionary
FEATURE_GROUPS = {
    'Identifier'       : ['instant'],
    'Time Feature'     : ['dteday', 'yr', 'mnth', 'hr', 'weekday', 'season'],
    'Calendar Feature' : ['holiday', 'workingday'],
    'Weather Feature'  : ['weathersit', 'temp', 'atemp', 'hum', 'windspeed'],
    'User Feature'     : ['casual', 'registered'],
    'Target'           : ['cnt'],
}

print('=' * 55)
print('KLASIFIKASI FEATURE SESUAI DATA DICTIONARY')
print('=' * 55)
for group, cols in FEATURE_GROUPS.items():
    print(f'\n[{group}]')
    for col in cols:
        print(f'  - {col} ({df[col].dtype})')

KLASIFIKASI FEATURE SESUAI DATA DICTIONARY

[Identifier]
  - instant (int64)

[Time Feature]
  - dteday (object)
  - yr (int64)
  - mnth (int64)
  - hr (int64)
  - weekday (int64)
  - season (int64)

[Calendar Feature]
  - holiday (int64)
  - workingday (int64)

[Weather Feature]
  - weathersit (int64)
  - temp (float64)
  - atemp (float64)
  - hum (float64)
  - windspeed (float64)

[User Feature]
  - casual (int64)
  - registered (int64)

[Target]
  - cnt (int64)


**Interpretasi:**

Dataset memiliki 6 kelompok feature:

| Kelompok | Jumlah Kolom | Digunakan pada Model |
|----------|-------------|----------------------|
| Identifier | 1 | ❌ Tidak |
| Time Feature | 6 | ✅ Ya |
| Calendar Feature | 2 | ✅ Ya |
| Weather Feature | 5 | ✅ Ya |
| User Feature | 2 | ❌ Tidak (data leakage) |
| Target | 1 | 🎯 Target Prediksi |

Feature `instant`, `casual`, dan `registered` tidak akan digunakan sebagai input model. `casual` dan `registered` dikecualikan karena keduanya merupakan penyusun langsung dari `cnt` sehingga penggunaannya akan menyebabkan **data leakage**.

### 5.4 Konfirmasi Target Prediksi

In [11]:
# Konfirmasi kolom target prediksi
TARGET_COL = 'cnt'

print(f'Target Prediksi : {TARGET_COL}')
print(f'Tipe Data       : {df[TARGET_COL].dtype}')
print(f'Nilai Min       : {df[TARGET_COL].min()}')
print(f'Nilai Max       : {df[TARGET_COL].max()}')
print(f'Nilai Rata-rata : {df[TARGET_COL].mean():.2f}')

# Verifikasi hubungan cnt = casual + registered
verifikasi = (df['cnt'] == df['casual'] + df['registered']).all()
print(f'\nVerifikasi cnt = casual + registered : {verifikasi}')

Target Prediksi : cnt
Tipe Data       : int64
Nilai Min       : 1
Nilai Max       : 977
Nilai Rata-rata : 189.46

Verifikasi cnt = casual + registered : True


**Interpretasi:**

- Kolom `cnt` merupakan target prediksi dengan nilai integer non-negatif.
- Persamaan `cnt = casual + registered` terverifikasi untuk seluruh baris, mengonfirmasi bahwa `casual` dan `registered` **tidak boleh digunakan** sebagai feature input model.

---

## 6. Missing Value Analysis

Memeriksa keberadaan nilai yang hilang (missing value) pada seluruh kolom dataset. Identifikasi missing value sangat penting untuk menentukan strategi data cleaning yang tepat.

In [12]:
# Analisis missing value
missing_count = df.isnull().sum()
missing_pct   = (df.isnull().sum() / len(df) * 100).round(4)

missing_df = pd.DataFrame({
    'Kolom'               : missing_count.index,
    'Jumlah Missing'      : missing_count.values,
    'Persentase Missing (%)': missing_pct.values
}).reset_index(drop=True)

print('=' * 55)
print('ANALISIS MISSING VALUE')
print('=' * 55)
print(missing_df.to_string(index=False))
print(f'\nTotal kolom dengan missing value : {(missing_count > 0).sum()}')

ANALISIS MISSING VALUE
     Kolom  Jumlah Missing  Persentase Missing (%)
   instant               0                  0.0000
    dteday               0                  0.0000
    season               0                  0.0000
        yr               0                  0.0000
      mnth               0                  0.0000
        hr               0                  0.0000
   holiday               0                  0.0000
   weekday               0                  0.0000
workingday               0                  0.0000
weathersit               0                  0.0000
      temp               0                  0.0000
     atemp               0                  0.0000
       hum               0                  0.0000
 windspeed               0                  0.0000
    casual               0                  0.0000
registered               0                  0.0000
       cnt               0                  0.0000

Total kolom dengan missing value : 0


**Interpretasi:**

Dataset **tidak memiliki missing value** pada seluruh 17 kolom. Kondisi ini sesuai dengan informasi pada Data Dictionary dan menunjukkan kualitas data yang baik sehingga tidak diperlukan strategi imputasi. Tahap Data Cleaning selanjutnya dapat difokuskan pada aspek lain seperti tipe data dan penanganan outlier jika diperlukan.

---

## 7. Duplicate Analysis

Memeriksa keberadaan baris duplikat pada dataset. Baris duplikat dapat mengakibatkan bias pada model jika tidak ditangani.

In [13]:
# Analisis duplikasi data
total_duplikat = df.duplicated().sum()
total_baris    = len(df)

print('=' * 45)
print('ANALISIS DUPLICATE DATA')
print('=' * 45)
print(f'Total baris               : {total_baris:,}')
print(f'Jumlah duplikat (seluruh kolom) : {total_duplikat}')
print(f'Persentase duplikat       : {total_duplikat / total_baris * 100:.4f}%')

ANALISIS DUPLICATE DATA
Total baris               : 17,379
Jumlah duplikat (seluruh kolom) : 0
Persentase duplikat       : 0.0000%


In [14]:
# Memeriksa duplikat berdasarkan kolom kunci time series (dteday + hr)
duplikat_kunci = df.duplicated(subset=['dteday', 'hr']).sum()

print('=' * 45)
print('DUPLIKAT BERDASARKAN KUNCI TIME SERIES')
print('Kunci: dteday + hr (tanggal + jam)')
print('=' * 45)
print(f'Jumlah duplikat kunci : {duplikat_kunci}')

DUPLIKAT BERDASARKAN KUNCI TIME SERIES
Kunci: dteday + hr (tanggal + jam)
Jumlah duplikat kunci : 0


**Interpretasi:**

Dataset **tidak mengandung baris duplikat**, baik secara keseluruhan maupun berdasarkan kombinasi kunci time series `dteday + hr`. Hal ini menunjukkan bahwa setiap jam memiliki satu observasi unik dan data sudah tertata dengan rapi. **Tidak diperlukan penghapusan duplikat.**

---

## 8. Statistical Summary

Menampilkan statistik deskriptif untuk memahami distribusi nilai pada setiap kolom numerik.

### 8.1 Statistik Deskriptif Umum

In [15]:
# Statistik deskriptif seluruh kolom
df.describe().T

,count,mean,std,min,25%,50%,75%,max
instant,17379.0000,8690.0000,5017.0295,1.0000,4345.5000,8690.0000,13034.5000,17379.0000
season,17379.0000,2.5016,1.1069,1.0000,2.0000,3.0000,3.0000,4.0000
yr,17379.0000,0.5026,0.5000,0.0000,0.0000,1.0000,1.0000,1.0000
mnth,17379.0000,6.5378,3.4388,1.0000,4.0000,7.0000,10.0000,12.0000
hr,17379.0000,11.5468,6.9144,0.0000,6.0000,12.0000,18.0000,23.0000
holiday,17379.0000,0.0288,0.1672,0.0000,0.0000,0.0000,0.0000,1.0000
weekday,17379.0000,3.0037,2.0058,0.0000,1.0000,3.0000,5.0000,6.0000
workingday,17379.0000,0.6827,0.4654,0.0000,0.0000,1.0000,1.0000,1.0000
weathersit,17379.0000,1.4253,0.6394,1.0000,1.0000,1.0000,2.0000,4.0000
temp,17379.0000,0.4970,0.1926,0.0200,0.3400,0.5000,0.6600,1.0000


### 8.2 Statistik Tambahan (Median)

Menambahkan median yang tidak tersedia pada `describe()` bawaan.

In [16]:
# Kolom numerik yang relevan untuk dianalisis
kolom_numerik = ['temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered', 'cnt']

# Ringkasan statistik lengkap
stats = df[kolom_numerik].agg(['mean', 'median', 'min', 'max', 'std']).T
stats.columns = ['Mean', 'Median', 'Min', 'Max', 'Std Dev']
stats = stats.round(4)

print('=' * 65)
print('RINGKASAN STATISTIK KOLOM NUMERIK')
print('=' * 65)
print(stats.to_string())

RINGKASAN STATISTIK KOLOM NUMERIK
               Mean   Median    Min      Max  Std Dev
temp         0.4970   0.5000 0.0200   1.0000   0.1926
atemp        0.4758   0.4848 0.0000   1.0000   0.1719
hum          0.6272   0.6300 0.0000   1.0000   0.1929
windspeed    0.1901   0.1940 0.0000   0.8507   0.1223
casual      35.6762  17.0000 0.0000 367.0000  49.3050
registered 153.7869 115.0000 0.0000 886.0000 151.3573
cnt        189.4631 142.0000 1.0000 977.0000 181.3876


**Interpretasi:**

**Kolom Target (`cnt`):**
- Rata-rata penyewaan per jam adalah sekitar 189 sepeda.
- Nilai minimum 1 dan maksimum 977 menunjukkan variasi yang sangat tinggi.
- Standar deviasi yang besar (~181) mengindikasikan pola yang sangat dipengaruhi waktu dan kondisi.

**Kolom Cuaca:**
- `temp` dan `atemp` memiliki nilai yang berdekatan, menunjukkan korelasi tinggi antar keduanya.
- `hum` (kelembapan) rata-rata sekitar 0.627 dalam skala 0–1.
- `windspeed` relatif rendah dengan rata-rata sekitar 0.19.

**Kolom User:**
- `registered` memiliki nilai rata-rata jauh lebih tinggi dari `casual`, menunjukkan pengguna terdaftar mendominasi total penyewaan.

### 8.3 Statistik Kolom Kategorikal

In [17]:
# Statistik kolom kategorikal (nilai unik dan frekuensi)
kolom_kategorikal = ['season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit']

print('=' * 55)
print('RINGKASAN KOLOM KATEGORIKAL')
print('=' * 55)
for col in kolom_kategorikal:
    nilai_unik = sorted(df[col].unique())
    print(f'\n{col} | Nilai unik: {nilai_unik}')
    print(df[col].value_counts().sort_index().to_string())

RINGKASAN KOLOM KATEGORIKAL

season | Nilai unik: [np.int64(1), np.int64(2), np.int64(3), np.int64(4)]
season
1    4242
2    4409
3    4496
4    4232

yr | Nilai unik: [np.int64(0), np.int64(1)]
yr
0    8645
1    8734

mnth | Nilai unik: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12)]
mnth
1     1429
2     1341
3     1473
4     1437
5     1488
6     1440
7     1488
8     1475
9     1437
10    1451
11    1437
12    1483

hr | Nilai unik: [np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23)]
hr
0     726
1     724
2     715
3     697
4     697
5     717
6     725
7     727
8     727
9     727
10    727
11    7

**Interpretasi:**

- **`season`**: Data mencakup 4 musim dengan distribusi yang relatif merata.
- **`yr`**: Dua tahun data (0 = 2011, 1 = 2012).
- **`mnth`**: Data mencakup seluruh 12 bulan dalam setahun.
- **`hr`**: Data mencakup seluruh 24 jam (0–23).
- **`holiday`**: Mayoritas data adalah hari non-libur.
- **`weekday`**: Data terdistribusi merata di seluruh hari dalam seminggu.
- **`workingday`**: Sekitar 68% data merupakan hari kerja.
- **`weathersit`**: Mayoritas data berada pada kondisi cuaca cerah (nilai 1), dan tidak ada atau sedikit sekali data dengan cuaca ekstrem (nilai 4).

---

## 9. Validasi Data Dictionary

Membandingkan dataset dengan dokumentasi Data Dictionary untuk memastikan kesesuaian struktur, nama feature, tipe data, dan target prediksi.

### 9.1 Validasi Jumlah Feature

In [18]:
# Jumlah feature sesuai dokumentasi
EXPECTED_ROWS = 17379
EXPECTED_COLS = 17

print('=' * 50)
print('VALIDASI DIMENSI DATASET')
print('=' * 50)
print(f'Jumlah baris  — Dokumentasi : {EXPECTED_ROWS:,} | Aktual : {len(df):,} | '
      f'Status: {"✅ Sesuai" if len(df) == EXPECTED_ROWS else "❌ Tidak Sesuai"}')
print(f'Jumlah kolom  — Dokumentasi : {EXPECTED_COLS}   | Aktual : {len(df.columns)}   | '
      f'Status: {"✅ Sesuai" if len(df.columns) == EXPECTED_COLS else "❌ Tidak Sesuai"}')

VALIDASI DIMENSI DATASET
Jumlah baris  — Dokumentasi : 17,379 | Aktual : 17,379 | Status: ✅ Sesuai
Jumlah kolom  — Dokumentasi : 17   | Aktual : 17   | Status: ✅ Sesuai


### 9.2 Validasi Nama Kolom

In [19]:
# Nama kolom sesuai dokumentasi
EXPECTED_COLUMNS = [
    'instant', 'dteday', 'season', 'yr', 'mnth', 'hr',
    'holiday', 'weekday', 'workingday', 'weathersit',
    'temp', 'atemp', 'hum', 'windspeed',
    'casual', 'registered', 'cnt'
]

print('=' * 55)
print('VALIDASI NAMA KOLOM')
print('=' * 55)

kolom_aktual = list(df.columns)
for col in EXPECTED_COLUMNS:
    status = '✅ Ada' if col in kolom_aktual else '❌ Tidak Ada'
    print(f'  {status} : {col}')

kolom_tidak_terdokumentasi = [c for c in kolom_aktual if c not in EXPECTED_COLUMNS]
if kolom_tidak_terdokumentasi:
    print(f'\n⚠ Kolom tidak terdokumentasi: {kolom_tidak_terdokumentasi}')
else:
    print(f'\nSeluruh kolom sesuai dengan dokumentasi Data Dictionary.')

VALIDASI NAMA KOLOM
  ✅ Ada : instant
  ✅ Ada : dteday
  ✅ Ada : season
  ✅ Ada : yr
  ✅ Ada : mnth
  ✅ Ada : hr
  ✅ Ada : holiday
  ✅ Ada : weekday
  ✅ Ada : workingday
  ✅ Ada : weathersit
  ✅ Ada : temp
  ✅ Ada : atemp
  ✅ Ada : hum
  ✅ Ada : windspeed
  ✅ Ada : casual
  ✅ Ada : registered
  ✅ Ada : cnt

Seluruh kolom sesuai dengan dokumentasi Data Dictionary.


### 9.3 Validasi Tipe Data

In [20]:
# Tipe data sesuai dokumentasi
EXPECTED_DTYPES = {
    'instant'   : 'int64',
    'dteday'    : 'object',   # Perlu konversi ke datetime pada preprocessing
    'season'    : 'int64',
    'yr'        : 'int64',
    'mnth'      : 'int64',
    'hr'        : 'int64',
    'holiday'   : 'int64',
    'weekday'   : 'int64',
    'workingday': 'int64',
    'weathersit': 'int64',
    'temp'      : 'float64',
    'atemp'     : 'float64',
    'hum'       : 'float64',
    'windspeed' : 'float64',
    'casual'    : 'int64',
    'registered': 'int64',
    'cnt'       : 'int64',
}

print('=' * 60)
print('VALIDASI TIPE DATA')
print('=' * 60)
for col, expected_type in EXPECTED_DTYPES.items():
    actual_type = str(df[col].dtype)
    status = '✅ Sesuai' if actual_type == expected_type else f'⚠ Berbeda ({actual_type})'
    print(f'  {col:<12} | Ekspektasi: {expected_type:<8} | Status: {status}')

VALIDASI TIPE DATA
  instant      | Ekspektasi: int64    | Status: ✅ Sesuai
  dteday       | Ekspektasi: object   | Status: ✅ Sesuai
  season       | Ekspektasi: int64    | Status: ✅ Sesuai
  yr           | Ekspektasi: int64    | Status: ✅ Sesuai
  mnth         | Ekspektasi: int64    | Status: ✅ Sesuai
  hr           | Ekspektasi: int64    | Status: ✅ Sesuai
  holiday      | Ekspektasi: int64    | Status: ✅ Sesuai
  weekday      | Ekspektasi: int64    | Status: ✅ Sesuai
  workingday   | Ekspektasi: int64    | Status: ✅ Sesuai
  weathersit   | Ekspektasi: int64    | Status: ✅ Sesuai
  temp         | Ekspektasi: float64  | Status: ✅ Sesuai
  atemp        | Ekspektasi: float64  | Status: ✅ Sesuai
  hum          | Ekspektasi: float64  | Status: ✅ Sesuai
  windspeed    | Ekspektasi: float64  | Status: ✅ Sesuai
  casual       | Ekspektasi: int64    | Status: ✅ Sesuai
  registered   | Ekspektasi: int64    | Status: ✅ Sesuai
  cnt          | Ekspektasi: int64    | Status: ✅ Sesuai


### 9.4 Validasi Target Prediksi

In [21]:
# Validasi target prediksi
EXPECTED_TARGET = 'cnt'

print('=' * 45)
print('VALIDASI TARGET PREDIKSI')
print('=' * 45)
status = '✅ Sesuai' if EXPECTED_TARGET in df.columns else '❌ Tidak Ditemukan'
print(f'Target Prediksi — Dokumentasi : {EXPECTED_TARGET} | Status: {status}')

VALIDASI TARGET PREDIKSI
Target Prediksi — Dokumentasi : cnt | Status: ✅ Sesuai


### 9.5 Validasi Nilai Unik Kolom Kategorikal

In [22]:
# Validasi nilai unik kolom kategorikal sesuai dokumentasi
EXPECTED_CATEGORIES = {
    'season'    : [1, 2, 3, 4],
    'yr'        : [0, 1],
    'mnth'      : list(range(1, 13)),
    'hr'        : list(range(0, 24)),
    'holiday'   : [0, 1],
    'weekday'   : [0, 1, 2, 3, 4, 5, 6],
    'workingday': [0, 1],
    'weathersit': [1, 2, 3, 4],
}

print('=' * 60)
print('VALIDASI NILAI UNIK KOLOM KATEGORIKAL')
print('=' * 60)
for col, expected_vals in EXPECTED_CATEGORIES.items():
    actual_vals = sorted(df[col].unique().tolist())
    sesuai = all(v in expected_vals for v in actual_vals)
    status = '✅ Sesuai' if sesuai else '⚠ Ada nilai di luar dokumentasi'
    print(f'  {col:<12} | Aktual: {actual_vals} | Status: {status}')

VALIDASI NILAI UNIK KOLOM KATEGORIKAL
  season       | Aktual: [1, 2, 3, 4] | Status: ✅ Sesuai
  yr           | Aktual: [0, 1] | Status: ✅ Sesuai
  mnth         | Aktual: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] | Status: ✅ Sesuai
  hr           | Aktual: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23] | Status: ✅ Sesuai
  holiday      | Aktual: [0, 1] | Status: ✅ Sesuai
  weekday      | Aktual: [0, 1, 2, 3, 4, 5, 6] | Status: ✅ Sesuai
  workingday   | Aktual: [0, 1] | Status: ✅ Sesuai
  weathersit   | Aktual: [1, 2, 3, 4] | Status: ✅ Sesuai


**Interpretasi:**

Hasil validasi menunjukkan:

- ✅ Jumlah baris dan kolom sesuai dokumentasi.
- ✅ Seluruh nama kolom terdokumentasi dengan benar.
- ✅ Tipe data sudah sesuai. Catatan: `dteday` bertipe `object` (string) dan akan dikonversi ke `datetime` pada tahap Data Cleaning.
- ✅ Target prediksi `cnt` terkonfirmasi.
- ✅ Nilai unik pada kolom kategorikal sesuai dengan mapping pada Data Dictionary. Jika `weathersit` hanya memiliki nilai 1–3, artinya kondisi cuaca ekstrem (nilai 4) sangat jarang atau tidak terjadi selama periode data.

**Catatan untuk Data Cleaning:**
- Kolom `dteday` perlu dikonversi dari string ke tipe `datetime`.
- Kolom `season`, `yr`, `hr`, `holiday`, `weekday`, `workingday`, dan `weathersit` secara semantik merupakan kategorikal meskipun bertipe integer.

---

## 10. Initial Findings

Ringkasan hasil Data Understanding berdasarkan seluruh pemeriksaan yang telah dilakukan.

### 10.1 Kualitas Dataset

- ✅ **Dataset bersih dan lengkap.** Tidak ditemukan missing value pada seluruh 17 kolom.
- ✅ **Tidak ada duplikasi data.** Setiap kombinasi `dteday + hr` adalah unik.
- ✅ **Struktur dataset sesuai dokumentasi.** Jumlah baris (17.379), jumlah kolom (17), nama kolom, dan target prediksi (`cnt`) semuanya tervalidasi.
- ✅ **Nilai kategorikal valid.** Tidak ditemukan nilai di luar rentang yang terdokumentasi.

---

### 10.2 Temuan Struktural

- Dataset merupakan **Multivariate Time Series** dengan granularitas per jam.
- Rentang waktu: **1 Januari 2011 – 31 Desember 2012** (2 tahun penuh).
- Kolom `dteday` bertipe `object` dan **harus dikonversi** ke `datetime` pada tahap Data Cleaning.
- Kolom `season`, `yr`, `mnth`, `hr`, `holiday`, `weekday`, `workingday`, dan `weathersit` **bersifat kategorikal** meskipun bertipe integer dalam dataset.

---

### 10.3 Temuan pada Target Prediksi

- Target prediksi adalah kolom `cnt` (total penyewaan per jam).
- Nilai `cnt` berkisar antara **1 hingga 977** dengan rata-rata sekitar **189**.
- Standar deviasi yang tinggi (~181) mengindikasikan **variasi yang besar** — analisis lebih lanjut pada EDA diperlukan untuk memahami pola distribusi.
- Hubungan `cnt = casual + registered` terverifikasi untuk **seluruh baris** data.

---

### 10.4 Catatan Data Leakage

- ⚠️ Kolom `casual` dan `registered` **tidak boleh digunakan** sebagai feature input karena merupakan penyusun langsung target `cnt`.
- ⚠️ Kolom `instant` tidak memiliki nilai prediktif dan akan dihapus sebelum modeling.

---

### 10.5 Kesiapan Dataset

Dataset **siap untuk Exploratory Data Analysis (EDA)** tanpa memerlukan perbaikan kualitas data yang mendesak. Persiapan minimal yang diperlukan sebelum modeling adalah:

1. Konversi tipe data `dteday` → `datetime`.
2. Encoding kolom kategorikal jika diperlukan (sesuai metode yang dipilih).
3. Penghapusan kolom `instant`, `casual`, dan `registered` sebelum modeling.
4. Analisis dan penanganan outlier (akan dilakukan pada EDA).

---

### 10.6 Potensi Feature Engineering

Berdasarkan pemeriksaan awal, beberapa potensi feature engineering yang dapat dilakukan:

- Ekstraksi dari `dteday`: minggu ke-, kuartal, awal/akhir bulan.
- Representasi siklus untuk `hr`, `mnth`, `weekday` menggunakan fungsi sinus/kosinus.
- Lag features dari `cnt` untuk menangkap pola historis.
- Interaksi antara `hr` dan `workingday` untuk membedakan jam sibuk.

---

## 11. Next Step

---

### Eksperimen Berikutnya

| Informasi | Detail |
|-----------|--------|
| **ID** | EXP-002 |
| **Nama** | Exploratory Data Analysis (EDA) |
| **Notebook** | `02_eda.ipynb` |

**Tujuan:**

> Mengeksplorasi pola, distribusi, hubungan antar feature, dan karakteristik target sebelum dilakukan preprocessing dan pemodelan.

**Aktivitas yang akan dilakukan pada EXP-002:**

- Analisis distribusi target `cnt`
- Analisis pola harian, mingguan, dan musiman
- Visualisasi hubungan antar feature
- Analisis korelasi antar variabel
- Identifikasi outlier
- Formulasi hipotesis untuk feature engineering

---

*Notebook EXP-001 Data Understanding — Selesai.*